instrumentation/httpurlconnection/testing/src/androidTest/java/io/opentelemetry/instrumentation/library/httpurlconnection/InstrumentationTest.java

```diff
@@ -0,0 +1,135 @@
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+
+public class InstrumentationTest {
+    private static final InMemorySpanExporter inMemorySpanExporter = InMemorySpanExporter.create();
+
+    @Before
+    public void setUp() {
+        OpenTelemetryTestUtils.setUpSpanExporter(inMemorySpanExporter);
+    }
+
+    @After
+    public void tearDown() {
+        inMemorySpanExporter.reset();
+    }
+
+    @Test
+    public void testHttpUrlConnectionGetRequest_ShouldBeTraced() {
+        HttpUrlConnectionTestUtil.executeGet("http://httpbin.org/get");
+        assertEquals(1, inMemorySpanExporter.getFinishedSpanItems().size());
+    }
+
+    @Test
+    public void testHttpUrlConnectionPostRequest_ShouldBeTraced() {
+        HttpUrlConnectionTestUtil.post("http://httpbin.org/post");
+        assertEquals(1, inMemorySpanExporter.getFinishedSpanItems().size());
+    }
+
+    @Test
+    public void
+            testHttpUrlConnectionGetRequest_WhenNoStreamFetchedAndNoDisconnectCalled_ShouldNotBeTraced() {
+        HttpUrlConnectionTestUtil.executeCustomGet("http://httpbin.org/get", false, false);
+        assertEquals(0, inMemorySpanExporter.getFinishedSpanItems().size());
+    }
+
+    @Test
+    public void
+            testHttpUrlConnectionGetRequest_WhenNoStreamFetchedButDisconnectCalled_ShouldBeTraced() {
+        HttpUrlConnectionTestUtil.executeCustomGet("http://httpbin.org/get", false, true);
+        assertEquals(1, inMemorySpanExporter.getFinishedSpanItems().size());
+    }
+
+    @Test
+    public void
+            testHttpUrlConnectionGetRequest_WhenNoStreamFetchedAndNoDisconnectCalledButHarvesterScheduled_ShouldBeTraced() {
+        HttpUrlConnectionTestUtil.executeCustomGet("http://httpbin.org/get", false, false);
+        ScheduledExecutorService harvester = scheduleHarvester();
+        try {
+            Thread.sleep(15000);
+            assertEquals(1, inMemorySpanExporter.getFinishedSpanItems().size());
+        } catch (InterruptedException e) {
+            Assert.fail("Test could not be completed as thread was interrupted while sleeping.");
+        } finally {
+            harvester.shutdown();
+        }
+    }
+
+    @Test
+    public void
+            testHttpUrlConnectionGetRequest_WhenFourConcurrentRequestsAreMade_AllShouldBeTraced() {
+        ExecutorService executor = Executors.newFixedThreadPool(4);
+        executor.submit(() -> HttpUrlConnectionTestUtil.executeGet("http://httpbin.org/get"));
+        executor.submit(() -> HttpUrlConnectionTestUtil.executeGet("http://google.com"));
+        executor.submit(() -> HttpUrlConnectionTestUtil.executeGet("http://android.com"));
+        executor.submit(() -> HttpUrlConnectionTestUtil.executeGet("http://httpbin.org/headers"));
+
+        executor.shutdown();
```

Comment:
I think this or `shutdownNow` should be in the finally block

Reply:
The shutdown here is a smooth shutdown called before executor.awaitTermination, that allows tasks to complete before shutting down the executor. And the executor.shutdownNow() in catch statement handles the scenario for shutting down executor if an exception occurs in awaitTermination call. 

That being said, the existing code does not initiate shutdown if exception were to occur in the four executor.submit calls. I'll be re-arranging the code to account for that as well.

Reply:
tbh, it won't make any difference
https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ExecutorService.html#shutdown--
`shutdown` just does not allow new submits, for this test specifically unless I misunderstood something, does nothing (because you don't submit any more tasks nor any other thread does).
`shutdown` and `shutdownNow` are more for cleaning resources/avoiding memory leaks in general, in tests since GC cleans up everything after the test is finished, won't make much difference, but of course, it's still a good practice.

Reply:
The javadoc descriptions around these are definitely confusing! We can't call awaitTermination without calling a shutdown first, the awaitTermination java description covers that:
- https://docs.oracle.com/javase%2F8%2Fdocs%2Fapi%2F%2F/java/util/concurrent/ExecutorService.html#awaitTermination-long-java.util.concurrent.TimeUnit-

- https://howtodoinjava.com/java/multi-threading/executorservice-shutdown/#:~:text=The%20awaitTermination(long%20timeout%2C%20TimeUnit,after%20a%20shutdown()%20request.

Reply:
true, it's been a while since I last used such methods.